Add console logging message to missing or late showNotification calls in service worker

youennf opened the pull request:
#### 9eb7504c1b928c5e5da74a591fc42dc5aef58f46
<pre>
Add console logging message to missing or late showNotification calls in service worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=254811">https://bugs.webkit.org/show_bug.cgi?id=254811</a>
rdar://problem/107471641

Reviewed by Ben Nham.

Add console log messages in case no showNotification is called during the lifetime of the push event.
Add console log messages in case a showNotification is called slightly too late after the end of the push event.

Add test API to be able to recover these console messages for WTR and related internals API.
Covered by added test.

* LayoutTests/http/wpt/service-workers/push-console-messages-no-showNotification.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-no-showNotification.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-async.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-async.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-late.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-late.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-sync.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-sync.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-worker.js: Added.
(async doPush):
(async doTest):
* LayoutTests/http/wpt/service-workers/push-console-messages.js: Added.
(promise_test.async test):
(async doTest):
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-no-showNotification.https-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-showNotification-async.https-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-showNotification-late.https-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-showNotification-sync.https-expected.txt: Added.
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
(WebCore::ServiceWorkerInternals::enableConsoleMessageReporting):
(WebCore::logReportedConsoleMessage):
* Source/WebCore/testing/ServiceWorkerInternals.h:
* Source/WebCore/testing/ServiceWorkerInternals.idl:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::dispatchPushEvent):
(WebCore::ServiceWorkerGlobalScope::didFirePushEventRecently const):
(WebCore::ServiceWorkerGlobalScope::addConsoleMessage):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFirePushEvent):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::reportConsoleMessage):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.messages.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::reportConsoleMessage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::reportServiceWorkerConsoleMessage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::reportServiceWorkerConsoleMessage):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::reportConsoleMessage):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
(-[ServiceWorkerOpenWindowWebsiteDataStoreDelegate websiteDataStore:openWindow:fromServiceWorkerOrigin:completionHandler:]):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::receivedServiceWorkerConsoleMessage):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate websiteDataStore:reportServiceWorkerConsoleMessage:]):

Canonical link: <a href="https://commits.webkit.org/262584@main">https://commits.webkit.org/262584@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1bd089c955cf6245b10912d383d823ef37cbde28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1944 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1975 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2037 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2871 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2032 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1920 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2021 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1799 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1963 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1751 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1741 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2717 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1724 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1641 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1788 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2889 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1800 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1727 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1735 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/479 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1891 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -319,6 +316,12 @@ void ServiceWorkerRegistration::showNotification(ScriptExecutionContext& context
         }
+
+        if (!serviceWorkerGlobalScope->hasPendingSilentPushEvent() && serviceWorkerGlobalScope->didFirePushEventRecently())
+            serviceWorkerGlobalScope->addConsoleMessage(MessageSource::Storage, MessageLevel::Warning, "showNotification was called outside of push event lifetime"_s, 0);
```
Should this message also mention that you may want to pass the promised returned by `showNotification` to the push event's `waitUntil` to appropriately extend the lifetime of the push event?

This is wordy, but I think we need something like: "This service worker showed a notification outside of the lifetime of a recent push event. This may cause the push event to be spuriously marked as a silent push. To fix this, consider passing the promise returned by showNotification to the push event's waitUntil. This will extend the lifetime of the push event to encompass the lifetime of the notification."

An alternative is to put the wordy explanation on the web somewhere, and have this error message link to the wordy explanation. I've seen Chrome do that in console messages but I'm not sure if we do that anywhere.

youennf replied:
OK, I'll add a mention to PushEvent.waitUntil, trying to stick with a short version.

bnham commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -303,9 +303,6 @@ void ServiceWorkerRegistration::showNotification(ScriptExecutionContext& context
         return;
     }
 
-    if (context.isServiceWorkerGlobalScope())
-        downcast<ServiceWorkerGlobalScope>(context).setHasPendingSilentPushEvent(false);
-
```
I think we originally put this check here defensively so that the web developer gets credit for trying to show a notification even if creating the notification creation fails for some reason. But looking through `Notification::createForServiceWorker`, this basically shouldn't happen except in edge cases (e.g. if the web dev tries to pass in a too-large `data` property into the Notification constructor). So I guess it's okay to move this down a few lines to where you've moved it.

youennf replied:
Right, I think we should move it, see for instance rdar://107461729

bnham reviewed (approved):
LGTM. I do wonder if we should try to be more specific in the error message and provide guidance about calling waitUntil on the push event,since that seems to be a common issue devs are running into.

bnham commented on Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
```diff
@@ -246,6 +246,9 @@ void ServiceWorkerThread::queueTaskToFirePushEvent(std::optional<Vector<uint8_t>
             bool showedNotification = !serviceWorkerGlobalScope->hasPendingSilentPushEvent();
+            serviceWorkerGlobalScope->setHasPendingSilentPushEvent(false);
+            if (!showedNotification)
+                serviceWorkerGlobalScope->addConsoleMessage(MessageSource::Storage, MessageLevel::Error, "Push event ended without showing any notification"_s, 0);
```
I wonder if we should be explicit about the consequences here. Like "Push event ended without showing any notification. After three such silent push events, the browser will remove your push subscription."

youennf replied:
I'll mention the risk of the push subscription removal, but not too precise (I do not want to change the message if we change the rule).

youennf committed:
Add console logging message to missing or late showNotification calls in service worker
https://bugs.webkit.org/show_bug.cgi?id=254811
rdar://problem/107471641

Reviewed by Ben Nham.

Add console log messages in case no showNotification is called during the lifetime of the push event.
Add console log messages in case a showNotification is called slightly too late after the end of the push event.

Add test API to be able to recover these console messages for WTR and related internals API.
Covered by added test.

* LayoutTests/http/wpt/service-workers/push-console-messages-no-showNotification.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-no-showNotification.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-async.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-async.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-late.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-late.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-sync.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-showNotification-sync.https.html: Added.
* LayoutTests/http/wpt/service-workers/push-console-messages-worker.js: Added.
(async doPush):
(async doTest):
* LayoutTests/http/wpt/service-workers/push-console-messages.js: Added.
(promise_test.async test):
(async doTest):
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-no-showNotification.https-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-showNotification-async.https-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-showNotification-late.https-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/service-workers/push-console-messages-showNotification-sync.https-expected.txt: Added.
* Source/WebCore/testing/ServiceWorkerInternals.cpp:
(WebCore::ServiceWorkerInternals::enableConsoleMessageReporting):
(WebCore::logReportedConsoleMessage):
* Source/WebCore/testing/ServiceWorkerInternals.h:
* Source/WebCore/testing/ServiceWorkerInternals.idl:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::dispatchPushEvent):
(WebCore::ServiceWorkerGlobalScope::didFirePushEventRecently const):
(WebCore::ServiceWorkerGlobalScope::addConsoleMessage):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFirePushEvent):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::reportConsoleMessage):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.messages.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::reportConsoleMessage):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::reportServiceWorkerConsoleMessage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::reportServiceWorkerConsoleMessage):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::reportConsoleMessage):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
(-[ServiceWorkerOpenWindowWebsiteDataStoreDelegate websiteDataStore:openWindow:fromServiceWorkerOrigin:completionHandler:]):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::receivedServiceWorkerConsoleMessage):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate websiteDataStore:reportServiceWorkerConsoleMessage:]):

Canonical link: https://commits.webkit.org/262584@main
